src/libstore/builtins.hh

```diff
@@ -4,6 +4,6 @@
 
 namespace nix {
 
-void builtinFetchurl(const BasicDerivation & drv, const std::string & netrcData);
+void builtinFetchurl(const BasicDerivation & drv, const std::string & netrcData, std::function<std::string(const std::string &)> rewriteStrings);
```

Comment:
I think it would be cleaner to rewrite strings in 'drv' prior to calling the builtin. That way all (future) builtin derivations benefit.

Reply:
@edolstra agree, and I wanted to do that but I didn't feel comfortable changing the working of the main codepath, since all the state machine and branching stuff in this file makes it very hard to understand (or feel confident that you really grok it).

Reply:
My inclination would be to remove `env` altogether (it's only used in a couple of places and is just another value that can get out of sync), and add the default environment values (as well as rewritten paths) to `drv->env`.